pkg/providers/vsphere/vmprovider_vm_utils.go

```diff
@@ -64,52 +65,104 @@ func GetVirtualMachineImageSpecAndStatus(
-			reason, msg := "NotFound", fmt.Sprintf("Failed to get the VM's image: %s", key.Name)
-			conditions.MarkFalse(vmCtx.VM, vmopv1.VirtualMachineConditionImageReady, reason, msg)
-			return nil, nil, nil, fmt.Errorf("%s: %w", msg, err)
+	switch vmCtx.VM.Spec.Image.Kind {
+	case "VirtualMachineImage":
+		var img vmopv1.VirtualMachineImage
+		if objErr = k8sClient.Get(vmCtx, key, &img); objErr == nil {
+			obj, spec, status = &img, &img.Spec, &img.Status
 		}
 
-		// Ensure the GVK for the object is synced back into the object since
-		// the object's APIVersion and Kind fields may be used later.
-		if err := kube.SyncGVKToObject(clusterVMImage, k8sClient.Scheme()); err != nil {
-			return nil, nil, nil, err
+	case "ClusterVirtualMachineImage":
+		key.Namespace = ""
+		var img vmopv1.ClusterVirtualMachineImage
+		if objErr = k8sClient.Get(vmCtx, key, &img); objErr == nil {
+			obj, spec, status = &img, &img.Spec, &img.Status
 		}
+	case "":
+		// This is only possible IFF VirtualMachine API resources created at a
+		// schema version prior to spec.image were not yet deployed when VM Op
+		// as upgraded to the schema version with spec.image. This means the
+		// create webhook did not run for those resources, populating their
+		// spec.image from spec.imageName. Therefore, to be safe, this case
+		// does what the webhook does -- looks up the image by both vmi and
+		// friendly name.
+		if img, err := vmopv1util.ResolveImageName(
+			vmCtx,
+			k8sClient,
+			vmCtx.VM.Namespace,
+			vmCtx.VM.Spec.Image.Name); err != nil {
+
+			objErr = err
+		} else {
+			switch timg := img.(type) {
+			case *vmopv1.VirtualMachineImage:
+				vmCtx.VM.Spec.Image = &vmopv1.VirtualMachineImageRef{
+					Kind: "VirtualMachineImage",
+					Name: timg.Name,
+				}
+				obj, spec, status = timg, &timg.Spec, &timg.Status
+			case *vmopv1.ClusterVirtualMachineImage:
+				vmCtx.VM.Spec.Image = &vmopv1.VirtualMachineImageRef{
+					Kind: "ClusterVirtualMachineImage",
+					Name: timg.Name,
+				}
+				obj, spec, status = timg, &timg.Spec, &timg.Status
+			}
+		}
+	default:
+		// This should never be possible in production as this function is
+		// only called in the create workflow, where spec.image is never nil.
+		// Still, it's better to test for this and set a condition in case the
+		// call stack for this function changes.
+		reason := "UnknownKind"
+		msg := vmCtx.VM.Spec.Image.Kind
+		conditions.MarkFalse(vmCtx.VM, vmopv1.VirtualMachineConditionImageReady, reason, msg)
+		return nil, nil, nil, fmt.Errorf("%s: %s", reason, msg)
+	}
 
-		obj, spec, status = clusterVMImage, &clusterVMImage.Spec, &clusterVMImage.Status
-	} else {
-		// Ensure the GVK for the object is synced back into the object since
-		// the object's APIVersion and Kind fields may be used later.
-		if err := kube.SyncGVKToObject(vmImage, k8sClient.Scheme()); err != nil {
-			return nil, nil, nil, err
+	if objErr != nil {
+		reason := "Unknown"
```

Comment:
I did it a while go as a quick hack until we do something better but there is a `errToConditionReasonAndMessage()` in this file that's unused elsewhere in similar situations.

Reply:
I will look.

Reply:
I am not going to use this right now as there are reasons not in the function that I would like to return.